aggie-coding-club/Rev-Registration
# Transition and memoize course cards

## Description

Users have commented that it feels jerky for cards to suddenly collapse and expand. Using Material-UI's `Collapse` component, I have added a 300ms transition between these states. I think `Collapse` probably uses some secret wizardry where they keep a pre-rendered copy of the collapsed card on hand so that they can rapidly add it back to the DOM during expansion.

Although `Collapse` makes it fast to expand and collapse the cards, it has no effect on switching between Basic and Section modes. I tried applying `useMemo` to `sectionSelect` to speed up this second transition, but to no avail. It appears that the time-consuming portion is not actually running the React code to determine what the `SectionSelect` should be, but to take that result and render it to the screen. (This was fairly surprising because of how inefficiently the `SectionSelect` component is written.) To eliminate the expensive step of rendering HTML elements to the DOM, I switched to using `display: none;` to hide the `customizationContent` that wasn't currently being used.

## Rationale

There is one sort of sketchy area where I manually calculate the height of the card (the `fixHeight` function), and this is liable to be accidentally broken in the future. Actually, it'll break as soon as Troy merges his Select All button. I'm not sure how I could add tests for it, though, so lmk if you have ideas.

Also, the part with the `MutationObserver` seems pretty hacky, too. The problem was that when the initially expanded card is rendered, it only has the loading circle, which is small enough to apply `expandedRowSmall`. So if I try to set the height of `sectionRows` while the `overflow: visible;` from `expandedRowSmall` is still active, it'll tell me I have all the space in the world. Instead, I waited until `CourseSelectColumn` changed the class to `expandedRow`, and then applied `fixHeight`. Waiting for the class to change meantt using a mutation observer. unless y'all have any better ideas.

## How to test

I'd like to try to get some automated tests for this, but JS DOM doesn't mock height and CSS rules IIRC, so I can't think of a way to do it.

For manual tests, open small classes like JAPN 201, medium classes like CSCE 314, and enormous classes like KINE 199. They should all collapse and expand smoothly no matter what order you open them in and regardless of how many times you switch between Basic and Section in between.

## Screenshots

|Before|After|
|--|--|
|![old](https://user-images.githubusercontent.com/10082177/99914302-14d96080-2cc2-11eb-986e-ad8c03f02749.gif)|![transitions](https://user-images.githubusercontent.com/10082177/99914237-97adeb80-2cc1-11eb-98d7-f5f9a037bd58.gif)|

## Related tasks

Closes #415 
Closes #387
